### PR TITLE
Prepare for Unicode 14

### PR DESCRIPTION
**Pull request by Manishearth**
This PR:

 - Adds Unicode 14 to all the requisite enums, etc
 - Copies Unicode 13's `ucd` and `emoji` folders.
 - Applies updated version/date headers to these files

(No other changes are made to the data files)

Some things missing from the [documentation](https://sites.google.com/site/unicodetools/home):
 - The `.bat` files are no longer generated and I couldn't figure out how to do it. Instead I followed the instructions [here](https://sites.google.com/site/unicodetools/inputdata) with the Python script
 - The `extra` and `cldr` folders do not exist under `unicodetools/data/ucd/14.0.0`, but they do under `Generated`. I only updated files which existed
 - Blocks.txt is under `unicodetools/data/ucd/14.0.0` in unicodetools, and `Generated/UCD/d1/extra` in Generated. Copying it out is not enough; the Generated Blocks.txt uses short names for blocks, whereas the unicodetools one uses long ones, and UCD asks for an additional file for the mapping. For now I just patched the header on Blocks.txt, but it seems like we should be generating `d1/Blocks.txt` as well with the long codes.

I can upload the `d1` files if you tell me when and where.

cc @markusicu @Ken-Whistler

### CODE REVIEW

**Review by macchiati — COMMENTED** [review]
What I usually do is diff them against the previous version, to check that the data looks sensible. I can do that, but probably later during the week.

One quick note is that the emoji files are showing version 13.0, which will need fixing. Not your problem since you were going by the instructions (and they were clearly incomplete)

**Comment by Manishearth** [comment]
> One quick note is that the emoji files are showing version 13.0, which will need fixing. Not your problem since you were going by the instructions (and they were clearly incomplete)

Should be easy enough to fix.

File: unicodetools/org/unicode/text/UCD/MakeUnicodeFiles.txt
Comment (markusicu) [inline unicodetools/org/unicode/text/UCD/MakeUnicodeFiles.txt:1]: What does the star do?

File: unicodetools/org/unicode/text/UCD/MakeUnicodeFiles.txt
Comment (markusicu) [inline unicodetools/org/unicode/text/UCD/MakeUnicodeFiles.txt:229]: September

http://www.unicode.org/versions/Unicode14.0.0/

File: unicodetools/org/unicode/text/utility/Utility.java
Comment (markusicu) [inline unicodetools/org/unicode/text/utility/Utility.java:829]: indentation? tabs?

**Review by markusicu — COMMENTED** [review]
Thanks, looks plausible for a start.

I always do a combination of following the notes on the unicodetools site plus “forensic programming”. I don't remember the details from one year to the next :-}

I see that you mostly copied Unicode 13 files as-is, but also put some Unicode 14 files in here. I think I used to start with a commit of just the old files in the new folders, for a more visible file history. Not sure if it makes much of a difference though. It *might* help git with the file history if the initial diffs for the new version are large enough to exceed git's threshold for whether the new version looks similar to the old one and record a "copy". Probably less important generally than in svn.

We will also need to copy the version-13 files for idna, security, and UCA, for a clean start there. Can be a separate PR.

As for generated files, they don't go directly back into the repo. We post the generated files to a location where KenW grabs them, looks them over, and if they are ok, he posts them in the Unicode Public folder. Then we grab all of the files from there again, put them into the repo's data folders, and run the tools again. Etc.

For example, here is where I posted files that I generated for Unicode 13: https://corp.unicode.org/~book/incoming/markus/u13/ plus (once UCD is settled) https://corp.unicode.org/~book/incoming/markus/uca13/

**Comment by markusicu** [comment]
> * The `extra` and `cldr` folders do not exist under `unicodetools/data/ucd/14.0.0`, but they do under `Generated`. I only updated files which existed

That's right. They are files for diagnosis or for use in CLDR and ICU, but don't go into the UCD/UCA/...

To be clear, I assume that what you put into the data input folders is straight from https://www.unicode.org/Public/14.0.0/ right? Do not check in things from the Generated output folder. The generated files first go for review, and we only pick them up after KenW has looked them over and copied them to /Public.

> * Blocks.txt is under `unicodetools/data/ucd/14.0.0` in unicodetools, and `Generated/UCD/d1/extra` in Generated. Copying it out is not enough; the Generated Blocks.txt uses short names for blocks, whereas the unicodetools one uses long ones, and UCD asks for an additional file for the mapping. For now I just patched the header on Blocks.txt, but it seems like we should be generating `d1/Blocks.txt` as well with the long codes.

I think we update Blocks.txt manually, post for Ken, wait for his approval and copy to /Public, and ony then copy into the data folder.

File: unicodetools/org/unicode/text/UCD/MakeUnicodeFiles.txt
Comment (Manishearth) [reply unicodetools/org/unicode/text/UCD/MakeUnicodeFiles.txt:1]: "Generate everything", i think this was in the docs somewhere, but I can revert it

**Comment by Manishearth** [comment]
> To be clear, I assume that what you put into the data input folders is straight from https://www.unicode.org/Public/14.0.0/ right? Do not check in things from the Generated output folder. The generated files first go for review, and we only pick them up after KenW has looked them over and copied them to /Public.

No, I copied them from Unicode 13, as listed in the docs. I can copy those files instead.

**Comment by Manishearth** [comment]
(once this is done I hope to update the docs with everything I've learned)

**Comment by markusicu** [comment]
> > To be clear, I assume that what you put into the data input folders is straight from https://www.unicode.org/Public/14.0.0/ right? Do not check in things from the Generated output folder. The generated files first go for review, and we only pick them up after KenW has looked them over and copied them to /Public.
> 
> No, I copied them from Unicode 13, as listed in the docs. I can copy those files instead.

You copied the ucd and emoji files from the data/.../13.0.0 folders to the data/.../14.0.0 folders, but you also then overwrote some files like unicodetools/data/ucd/14.0.0-Update/PropertyAliases.txt which has a version 14 header. What I am saying is that these files should only ever come from the /Public folder, never directly from the Generated output.

**Comment by Manishearth** [comment]
Ah. I'll make an additional commit copying in 14 files from Public

**Comment by Manishearth** [comment]
> but you also then overwrote some files like unicodetools/data/ucd/14.0.0-Update/PropertyAliases.txt which has a version 14 header

Oh, uh, not quite, this is the Unicode 13 file, I just added the Unicode 14 header to make the tools stop complaining about diffs 😄

**Comment by Manishearth** [comment]
- Copied in the generated files
 - Addressed inline comments
 - Added new scripts and blocks to the code. I did _not_ add them to the data files (e.g. PropertyValueAliases.txt for short codes), I expect they will be autogenerated there, but I can add them if necessary)

Currently stuck on

```
Exception in thread "main" java.lang.IllegalArgumentException: Failed to find Script value for 1E290
	at org.unicode.text.UCD.ToolUnicodePropertySource$ToolUnicodeProperty._getValue(ToolUnicodePropertySource.java:1653)
```

but 1E290 has an entry in Scripts.txt, so I'm not sure why it's not getting loaded

**Comment by Manishearth** [comment]
It seems to be generating again, after working through all the errors. Note for when I document stuff; we need to document "new scripts" and "new blocks".

**Comment by Manishearth** [comment]
Oh, another thing:

 - The emoji/, auxiliary/, and extracted/ folders on /Public are empty. I didn't touch them
 - Unihan.zip does not contain files that overlap with Unihan/, so I didn't update that

The build throws these non-blocking errors:

```
***Rules missing from SAMPLES for Line: [12.3, 13.04, 21.2, 27.01]
...
java.lang.IllegalArgumentException: Internal error: cjkAccountingNumeric doesn't contain : []
java.lang.IllegalArgumentException: Internal error: cjkOtherNumeric doesn't contain : []
java.lang.IllegalArgumentException: Internal error: cjkPrimaryNumeric doesn't contain : []
```

**Comment by markusicu** [comment]
> Oh, uh, not quite, this is the Unicode 13 file, I just added the Unicode 14 header to make the tools stop complaining about diffs 😄

I wouldn't make this kind of edit. The data files should be the old versions, or come from the /Public online folder, or have real edits for new data.

I *believe* that I used to make one commit (in svn) with just copies within the repo from the old-version data/ files to the new-version data/ files, and subsequent commits where I copied in /Public files and/or added new data.

> Copied in the generated files

This sounds wrong, as discussed...

> The emoji/, auxiliary/, and extracted/ folders on /Public are empty. I didn't touch them

They are empty until the files we generated get reviewed and posted.

> Unihan.zip does not contain files that overlap with Unihan/, so I didn't update that

The files in Unihan.zip are hard to work with. Hard to diff, and properties can move between files.
Instead, we re-munge them into one file per property.
Instructions are in the "Unihan" section of https://sites.google.com/site/unicodetools/inputdata
(We will need to change that slightly to talk about git instead of svn.)

----
FYI  I expect to post the last two ISO script codes tomorrow that we need for Unicode 14.
(I am the ISO 15924 registrar. There is a whole process for that.)

**Comment by Manishearth** [comment]
> I wouldn't make this kind of edit. The data files should be the old versions, or come from the /Public online folder, or have real edits for new data.


Yep, the files are now from /Public

**Comment by Manishearth** [comment]
Well, except for the emoji files, which don't exist and cause errors

**Comment by Manishearth** [comment]
Okay, the current state is that there are no more "copied from Unicode 13" files in `14.0.0-Update`. 

There are still files under `emoji/14.0` that are copied from Unicode 13, because the program fails without it. What should my next steps be here: should I be getting Ken to upload my generated files? Or getting this PR merged?

File: unicodetools/org/unicode/text/UCD/MakeUnicodeFiles.txt
Comment (markusicu) [inline unicodetools/org/unicode/text/UCD/MakeUnicodeFiles.txt:3]: We usually set this one higher than the highest dsomething that is published, so that there is no collision. Looks like we have a d7 now, so this wants to be 8 so that your generated files say d8.

File: unicodetools/org/unicode/text/UCD/ShortBlockNames.txt
Comment (markusicu) [inline unicodetools/org/unicode/text/UCD/ShortBlockNames.txt:321]:
```suggestion
Znamenny_Music ; Znamenny_Musical_Notation
```
see https://www.unicode.org/Public/14.0.0/ucd/PropertyValueAliases-14.0.0d2.txt

**Review by markusicu — COMMENTED** [review]
FYI  Mark has a competing PR now adding emoji 13.1. Let's see who gets in first vs. who gets to merge :-)

FYI  Next PR could update to https://www.unicode.org/Public/14.0.0/ucd/PropertyValueAliases-14.0.0d2.txt which adds new block and script names.

**Comment by macchiati** [comment]
Manish started first; the 13.1 update isn't as important...

Mark

**Comment by Manishearth** [comment]
Should I still pull in 13.1? 

Also, from the ucd-dev list it seems like a lot of data files have been updated: Should I be pulling in the latest files before we can merge this?

**Comment by macchiati** [comment]
13.1 was an emoji-sequences-only release, and the UCD files are identical
to 13.0.

 I just copied in 13.0 files because it makes it easier to build for the
UnicodeJsps (the tooling isn't set up quite right for having a delta like
that), so I think you can simply proceed with 14.0 as you were.

Mark

**Comment by markusicu** [comment]
> Should I still pull in 13.1?

I think yes. Just the four files from https://www.unicode.org/Public/emoji/13.1/

> Also, from the ucd-dev list it seems like a lot of data files have been updated: Should I be pulling in the latest files before we can merge this?

Possible, but I suggest to do that in the next PR.

**Comment by Manishearth** [comment]
> I think yes. Just the four files from https://www.unicode.org/Public/emoji/13.1/

Done (and addressed the other comments).

> Possible, but I suggest to do that in the next PR.

Will do, and then I suppose I should send the updated generated files to Ken?

**Comment by markusicu** [comment]
> > Possible, but I suggest to do that in the next PR.
> 
> Will do, and then I suppose I should send the updated generated files to Ken?

Sure. Best is to post the generated files somewhere and send to Ken with cc to ucd-dev.

We usually post to https://corp.unicode.org/~book/incoming/username -- do you have access to that? (Should be fine posting wherever is convenient.)

**Review by markusicu — APPROVED** [review]
looks plausible :-)

please squash & merge

**Comment by Manishearth** [comment]
Done! Going to pull in the latest data soon

> We usually post to https://corp.unicode.org/~book/incoming/username -- do you have access to that? (Should be fine posting wherever is convenient.)

I have the password for `book`, but I don't know how to upload to it: I'm able to log in to the web with `book` and get a dir listing, but not to FTP as listed in https://corp.unicode.org/~book/bookurls.txt . Do i need to get an account?

**Comment by Manishearth** [comment]
Posted https://github.com/unicode-org/unicodetools/pull/45 and https://corp.unicode.org/~book/incoming/manishearth/u14/d8/

**Comment by Manishearth** [comment]
I made some doc changes to update https://sites.google.com/site/unicodetools/home to the latest set of steps.

**Comment by macchiati** [comment]
Thanks!

Mark

**Pull request merged**
